xarray/tests/test_dataset.py

```diff
@@ -2558,6 +2558,21 @@ def test_rename_vars(self):
         with pytest.raises(ValueError):
             original.rename_vars(names_dict_bad)
 
+    def test_rename_to_dim_name(self):
+        # regression test for GH4107
+        coord_1 = xr.DataArray([1, 2], dims=["coord_1"], attrs={"attrs": True})
+        ds = xr.Dataset()
+        ds["a"] = xr.DataArray([1, 0], [coord_1])
+        obj = ds.reset_index("coord_1").rename({"coord_1_": "coord_1"})
+        assert_equal(ds, obj)
+
+        newds = (
+            ds.reset_index("coord_1")
+            .reset_coords("coord_1_")
+            .rename({"coord_1_": "coord_1"})
+        )
+        assert "coord_1" not in newds.data_vars
```

Comment:
Apparently we can create a dataset where a data variable and dimension share the same name. Should this be allowed? Should we raise an error instead of automatically promote `coord_1` to a coordinate variable?

```
<xarray.Dataset>
Dimensions:  (coord_1: 2)
Dimensions without coordinates: coord_1
Data variables:
    a        (coord_1) int64 1 0
    coord_1  (coord_1) int64 1 2
```

Reply:
now promotes to coords.